### ISSUE TITLE
Rule for anchor-is-valid contradicts anchor-has-content

### ISSUE THREAD

**Issue by svinkle** [evcohen/eslint-plugin-jsx-a11y#322]
I think I've spotted rule sets that contradict one another…

## The `anchor-is-valid` rule

In the description for the [`anchor-is-valid`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md) rule, it states:

> The HTML `<a>` element, with a valid `href` attribute, is formally defined as representing a **hyperlink**.

Then continues:

> An `<a>` element without an `href` attribute no longer functions as a hyperlink.

## The `anchor-has-content` rule

The **contradiction** takes place when reading over the [`anchor-has-content`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-has-content.md) rules, specifically what counts as a successful anchor link:

```html
<a>Anchor Content!</a>
<a><TextWrapper /><a>
<a dangerouslySetInnerHTML={{ __html: 'foo' }} />
```

I'd say these examples should also **fail** as they do not feature a valid `href` attribute. Without the `href` attribute, these anchors are not keyboard focusable and serve no real purpose (unless of course there's a click event listener attached.)

## Proposal for update

I'd like to propose the `anchor-has-content` rule be changed to succeed only when a valid `href` attribute is included with the anchor element.

---

I'd like to try and create a PR for this, but am unsure of how to setup a dev environment. Are there docs available on updating existing rules?

**Comment by ljharb** [comment]
I'm confused; why are these in conflict?

Without an href attribute, they're still anchors, because they're `<a>` tags, so anchor-has-content seems like it shouldn't be checking `href` - that's the job of `anchor-is-valid`.

**Comment by svinkle** [comment]
Perhaps "contradictory" is the incorrect term. I'm just confused that when I remove an `href` from an anchor there there's no warning that appears, either in the console or from eslint in my editor. I would assume this would be a default/core rule?

**Comment by svinkle** [comment]
I just noticed when running the `eslint src/` command, this includes an error about the `a` not having the `href` attribute. This is what I'm expecting to see elsewhere. Perhaps I don't have my `.eslintrc` configured correctly?

```
{
    "env": {
        "browser": true,
        "es6": true,
        "node": true
    },
    "parserOptions": {
        "ecmaVersion": 6,
        "sourceType": "module"
    },
    "plugins": [
        "jsx-a11y"
    ],
    "extends": [
        "plugin:jsx-a11y/recommended"
    ]
}
```

**Comment by ljharb** [comment]
Hmm, the recommended config [does enable both rules](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/src/index.js#L49-L50).

**Comment by svinkle** [comment]
I'll close this thread, but still strange how the expected error message is only output in the terminal when running `eslint` command. Browser console and editors are missing this important message. 🤔
